README.md
# deck-of-cards
Deck of Cards

## Principle of Least Surprise
A component of a system should behave in a way that most users will expect it to behave.
The behavior should not astonish or surprise users.
Formally: "If a necessary feature has a high astonishment factor, it may be necessary to redesign the feature."

### Build Test Scripts
You can build the included test scripts using cmake. 
There is a test script for:
- The Knuth shuffle algorithm [link](test/knuth_shuffle_test.cpp)
- Deck class, which shuffles and deals a Deck of PlayingCard objects [link](test/shuffle_deal_test.cpp)
- WarCardGame class, which runs an automated game of the War card game, written using the Deck and PlayingCard classes. [link](test/war_card_game_test.cpp)
```
mkdir build
cd build
cmake ..
cmake --build .
```

## [Explanation of process](explanation.txt)

## Classes

### Deck
file: [deck.h](include/deck.h)

### PlayingCard
files: [playing_card.h](include/playing_card.h), [playing_card.cpp](src/playing_card.cpp)

### WarCardGame
files: [war_card_games.h](include/war_card_games.h), [war_card_games.cpp](src/war_card_games.cpp)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2022-02-19
### Added
- Initial version (Deck, PlayingCard, WarCardGame classes)


test/shuffle_deal_test.cpp
/**
 * @file shuffle_deal_test.cpp
 *
 * @author Channing Ko-Madden
 *
 * @date 2/17/2022
 *
 * @brief Test script for creating a 52 card playing deck, shuffling it, and dealing the cards
 *
 */

#include "deck.h"
#include "playing_card.h"
#include <iostream>

int main(int argc, char** argv)
{
    doc::Deck<doc::PlayingCard> standardDeck(doc::buildStandardDeck());
    std::cout << "Unshuffled deck dealt out: \n";
    while (!standardDeck.empty())
    {
        std::cout << standardDeck.deal_card().str() << "\n";
    }


    std::cout << "Shuffled deck dealt out: \n";
    standardDeck.assign(doc::buildStandardDeck());
    standardDeck.shuffle();
    while (!standardDeck.empty())
    {
        std::cout << standardDeck.deal_card().str() << "\n";
    }

    std::cout << "Comparing shuffle & deal of two separate, newly created, 52 card decks: \n";
    doc::Deck<doc::PlayingCard> firstDeck(doc::buildStandardDeck());
    firstDeck.shuffle();
    doc::Deck<doc::PlayingCard> secondDeck(doc::buildStandardDeck());
    secondDeck.shuffle();
    while (!firstDeck.empty())
    {
        std::cout << firstDeck.deal_card().str() << "    " << secondDeck.deal_card().str() << "\n";
    }
    return 0;
}


test/knuth_shuffle_test.cpp
/**
 * @file knuth_shuffle_test.cpp
 *
 * @author Channing Ko-Madden
 *
 * @date 2/17/2022
 *
 * @brief Test script for the Knuth shuffle algorithm
 *
 * This script creates a vector with values from 1 to 10, and then uses the knuthShuffle function to
 * shuffle the vector multiple times, printing the result to the terminal, for the user to visually
 * see if it looks like a random shuffling is occurring.
 */

#include "deck.h"
#include <vector>
#include <numeric>
#include <random>
#include <iostream>
#include <sstream>

void printVector(const std::vector<int>& vect)
{
    std::stringstream outbuffer;
    for (const auto& value : vect)
    {
        outbuffer << value << ", ";
    }
    
    std::cout << outbuffer.str() << "\n";
}

int main(int argc, char** argv)
{

    std::vector<int> testVector(10);
    std::iota(testVector.begin(), testVector.end(), 1); // fill vector from 1 to size-1

    std::random_device randomSeeder;
    std::mt19937 mtGen(randomSeeder()); // randomly seed mt19937

    std::cout << "Initial vector: \n";
    printVector(testVector);
    std::cout << "Knuth shuffles: \n";
    for (int i = 0; i < 10; i ++)
    {
        doc::knuthShuffle(testVector, mtGen);
        printVector(testVector);
    }

    return 0;
}


test/war_card_game_test.cpp

/**
 * @file shuffle_deal_test.cpp
 *
 * @author Channing Ko-Madden
 *
 * @date 2/17/2022
 *
 * @brief Test script for the WarCardGame class, letting a game of war run automatically.
 *
 * Was also used to debug WAR phase edge cases.
 *
 */

#include "war_card_game.h"
#include "playing_card.h"
#include <vector>

int main(int argc, char** argv)
{
    // use custom deck to test WAR edge cases
    /*
    const doc::PlayingCard singleCard = doc::PlayingCard(
            doc::PlayingCard::Suite::Spades,
            doc::PlayingCard::Rank::Ace);
    const doc::PlayingCard secondCard = doc::PlayingCard(
            doc::PlayingCard::Suite::Spades,
            doc::PlayingCard::Rank::Two);
    std::vector<doc::PlayingCard> testCards(6, singleCard);
    testCards.push_back(secondCard);
    testCards.push_back(secondCard);
    doc::WarCardGame testGame(testCards);
    testGame.autoPlay();
    */

    // play a standard game
    doc::WarCardGame myGame;
    myGame.autoPlay();

    return 0;
}


include/deck.h
/**
 * @file deck.h
 *
 * @author Channing Ko-Madden
 *
 * @date 2/18/2022
 *
 * @brief Declare Deck class, that represents a deck of data of any size.
 */

#ifndef DECK_H
#define DECK_H

#include <vector>
#include <algorithm> // for std::swap
#include <random>

/**
 * @brief "Deck of Cards" namespace
 */
namespace doc
{

/**
 * @brief Implementation of Knuth Shuffle algorithm, also known as Fisher-Yates shuffle.
 * See https://en.wikipedia.org/wiki/Fisher-Yates_shuffle for background.
 * This algorithnm's time complexity is O(n).
 *
 * @param[in] data Data to shuffle
 * @param[in] randomGenerator 32-bit Mersenne Twister algorithm for generating random numbers.
 *  It is assumed this generator has been seeded appropriately.
 */
template <class T>
void knuthShuffle(std::vector<T>& data, std::mt19937& randomGenerator)
{
    std::uniform_int_distribution<unsigned long long> dist(0, data.size() - 1); // dist is range [a, size-1]
    for (auto& point : data)
    {
        std::swap(point, data.at(dist(randomGenerator)));
    }  
}

/**
 * @brief Exception thrown when an attempt to deal a card from an empty Deck occurs.
 */
class EmptyDeckException : public std::exception
{
    virtual const char * what() const throw ()
    {
        return("Deck is empty");
    }
};

/**
 * @class Deck "deck.h" "deck.h"
 * 
 * @brief Data is internally stored within an std::vector object. The "back" of the vector represents the top
 * of the deck, the "front" of the vector represents the bottom of the deck. A vector was chosen b/c it is efficient
 * at adding & removing from its end (amortized constant O(1), and drawing from the top of the deck is the
 * most common action done with a deck. It is also efficient for random access which is used by the shuffle method
 * to shuffle the deck.
 *
 * Other actions such as placing at the bottom of the deck, or placing randomly within the deck, are less common, and
 * I felt did not justify using a different container.
 *
 * T must meet the requirements of MoveAssignable and MoveConstructible
 */
template <class T>
class Deck
{
    public:

        /** @brief Construct empty deck 
         *
         * Initializes internal random number generator
         */
        Deck();

        /**
         * @brief Construct Deck with data
         *
         * Initializes internal random number generator
         * @param[in] data 
         */
        Deck(const std::vector<T>& data);

        /**
         * @brief No copy constructor to prevent multiple decks using the same seeded random number generator
         */
        Deck(const Deck&) = delete;

        /**
         * @brief No Copy assignment operator to prevent multiple decks using the same seeded random generator
         */
        Deck& operator=(const Deck&) = delete;

        /**
         * @brief Shuffle the deck, randomizing its contents
         */
        void shuffle();

        /**
         * @brief Returns if the deck is empty
         * @return True if the deck is empty, false otherwise */
        bool empty() const;

        /**
         * @brief Returns the size of deck
         * @return Size of the deck
         */
        size_t size() const;

        /**
         * @brief Returns a card from the top of the deck, removing it from the deck and reducing the size by one.
         *
         * Calling this method on an empty deck will throw an EmptyDeckException
         * @return Card at top of deck
         */
        T deal_card();

        /**
         * @brief Returns the contents of this deck, without clearing the deck.
         *
         * This method can be used to create a second Deck with the same contents of this Deck, but without
         * the same random number generator used to shuffle the deck.
         * @return Deck contents
         */
        std::vector<T> contents() const;

        /**
         * @brief Replaces the contents of the Deck
         * @param[in] data Data to replace contents.
         */
        void assign(const std::vector<T>& data);


    private:

        std::vector<T> mData; /**<@brief Deck's data */
        std::mt19937 mRandomGen; /**<@brief Use for shuffling */

};

// Definition below
template<class T>
Deck<T>::Deck()
{
    std::random_device randomSeeder;
    mRandomGen.seed(randomSeeder());
}

template<class T>
Deck<T>::Deck(const std::vector<T>& data) :
    mData(data)
{
    std::random_device randomSeeder;
    mRandomGen.seed(randomSeeder());
}

template<class T>
void Deck<T>::shuffle()
{
    knuthShuffle<T>(mData, mRandomGen);
}

template<class T>
bool Deck<T>::empty() const
{
    return mData.empty();
}

template<class T>
size_t Deck<T>::size() const
{
    return mData.size();
}

template<class T>
T Deck<T>::deal_card()
{
    if (mData.empty())
        throw EmptyDeckException();
    else
    {
        const T card = mData.back();
        mData.pop_back();
        return card;
    }
}

template<class T>
std::vector<T> Deck<T>::contents() const
{
    return mData;
}

template<class T>
void Deck<T>::assign(const std::vector<T>& data)
{
    mData = data;
}

} // namespace doc
#endif


src/playing_card.cpp
/*
 * @file playing_card.cpp
 *
 * @author Channing Ko-Madden
 *
 * @date 2/18/2022
 *
 * @brief Define PlayingCard class, that represents a French-suited playing card of a standard 52-card deck.
 *
 */

#include "playing_card.h"
#include <vector>
#include <sstream>
#include <type_traits>

namespace doc
{
    
// static
std::string PlayingCard::toString(const Suite suite)
{
    switch (suite)
    {
        case Suite::Clubs:
            return "Clubs";
        case Suite::Diamonds:
            return "Diamonds";
        case Suite::Hearts:
            return "Hearts";
        case Suite::Spades:
            return "Spades";
        default:
            return "";
    }
}

// static
std::string PlayingCard::toString(const Rank rank)
{
    switch (rank)
    {
        case Rank::Ace:
            return "A";
        case Rank::Two:
            return "2";
        case Rank::Three:
            return "3";
        case Rank::Four:
            return "4";
        case Rank::Five:
            return "5";
        case Rank::Six:
            return "6";
        case Rank::Seven:
            return "7";
        case Rank::Eight:
            return "8";
        case Rank::Nine:
            return "9";
        case Rank::Ten:
            return "10";
        case Rank::Jack:
            return "J";
        case Rank::Queen:
            return "Q";
        case Rank::King:
            return "K";
        default:
            return "";
    }
}

// static
int PlayingCard::toInt(const Rank rank)
{
    return static_cast<std::underlying_type<Rank>::type>(rank);
}

PlayingCard::PlayingCard(const Suite suite, const Rank rank) : mSuite(suite), mRank(rank)
{
}

PlayingCard::Suite PlayingCard::suite() const
{
    return mSuite;
}

PlayingCard::Rank PlayingCard::rank() const
{
    return mRank;
}

std::string PlayingCard::str() const
{
    std::stringstream outbuffer;
    outbuffer << PlayingCard::toString(mRank) << " " << PlayingCard::toString(mSuite);
    return outbuffer.str();
}

std::vector<PlayingCard> buildStandardDeck()
{
    const std::vector<PlayingCard::Suite> suites = {
        PlayingCard::Suite::Clubs,
        PlayingCard::Suite::Diamonds,
        PlayingCard::Suite::Hearts,
        PlayingCard::Suite::Spades};

    const std::vector<PlayingCard::Rank> ranks = {
        PlayingCard::Rank::King,
        PlayingCard::Rank::Queen,
        PlayingCard::Rank::Jack,
        PlayingCard::Rank::Ten,
        PlayingCard::Rank::Nine,
        PlayingCard::Rank::Eight,
        PlayingCard::Rank::Seven,
        PlayingCard::Rank::Six,
        PlayingCard::Rank::Five,
        PlayingCard::Rank::Four,
        PlayingCard::Rank::Three,
        PlayingCard::Rank::Two,
        PlayingCard::Rank::Ace};

    std::vector<PlayingCard> deck;
    deck.reserve(52);
    for (const auto& suite : suites)
    {
        for (const auto& rank : ranks)
        {
            deck.push_back(PlayingCard(suite, rank));
        }
    }

    return deck;
}

} // namespace doc


src/war_card_game.cpp
/*
 * @file war_card_game.cpp
 *
 * @author Channing Ko-Madden
 *
 * @date 2/18/2022
 *
 * @brief Define WarCardGame class, that simulates a game of the card game War, with the results
 * printed to the terminal.
 *
 */

#include "war_card_game.h"
#include "playing_card.h"
#include <iostream>
#include <vector>

namespace doc
{

WarCardGame::WarCardGame()
{
    initialize(doc::buildStandardDeck());
}

WarCardGame::WarCardGame(const std::vector<PlayingCard>& cards)
{
    if ((cards.size() % 2) == 1)
        initialize({cards.begin(), cards.end() - 1});
    else 
        initialize(cards);
}

void WarCardGame::playTurn()
{
    if (!gameOver())
    {
        mTurnCounter++;
        replenishDecks();

        const PlayingCard card1 = mP1Deck.deal_card();
        const PlayingCard card2 = mP2Deck.deal_card();
        std::cout << "Player 1 shows: " << card1.str() << "  Player 2 shows: " << card2.str() << std::endl;
        if (card1.rank() == card2.rank())
        {
            const int winner = war();
            if (winner == 1)
            {
                std::cout << "Player 1 Won the War\n";
                mP1WinPile.push_back(card1);
                mP1WinPile.push_back(card2);
                mP1WinPile.insert(mP1WinPile.end(), mP1WarCards.begin(), mP1WarCards.end());
                mP1WinPile.insert(mP1WinPile.end(), mP2WarCards.begin(), mP2WarCards.end());
            }
            else if (winner == 2)
            {
                std::cout << "Player 2 Won the War\n";
                mP2WinPile.push_back(card1);
                mP2WinPile.push_back(card2);
                mP2WinPile.insert(mP2WinPile.end(), mP2WarCards.begin(), mP2WarCards.end());
                mP2WinPile.insert(mP2WinPile.end(), mP1WarCards.begin(), mP1WarCards.end());
            }
            else // DRAW! Everyone gets their cards back
            {
                std::cout << " The War was a Draw ... bummer ...\n";
                mP1WinPile.push_back(card1);
                mP2WinPile.push_back(card2);
                mP1WinPile.insert(mP1WinPile.end(), mP1WarCards.begin(), mP1WarCards.end());
                mP2WinPile.insert(mP2WinPile.end(), mP2WarCards.begin(), mP2WarCards.end());
            }
            mP1WarCards.clear();
            mP2WarCards.clear();
        }
        else if (card1.rank() == PlayingCard::Rank::Ace)
        {
            //player 1 wins
            mP1WinPile.push_back(card1);
            mP1WinPile.push_back(card2);
        }
        else if (card2.rank() == PlayingCard::Rank::Ace)
        {
            // player 2 wins
            mP2WinPile.push_back(card1);
            mP2WinPile.push_back(card2);
        }
        else if (card1.rank() > card2.rank())
        {
            //player 1 wins
            mP1WinPile.push_back(card1);
            mP1WinPile.push_back(card2);
        }
        else
        {
            // player 2 wins
            mP2WinPile.push_back(card1);
            mP2WinPile.push_back(card2);
        }

    }
    else
    {
        std::cout << "Cannot play turn, game has already ended.\n";
    }
}

void WarCardGame::printScore() const
{
    const unsigned int p1Score = mP1Deck.size() + mP1WinPile.size();
    const unsigned int p2Score = mP2Deck.size() + mP2WinPile.size();

    std::cout << "Player One has " << p1Score << " cards left.   Player Two has " << p2Score <<
        " cards left.\n";
    if (p1Score == 0)
        std:: cout << "Player Two has won!\n";
    else if (p2Score == 0)
        std:: cout << "Player One has won!\n";
}

void WarCardGame::autoPlay()
{
    printScore();
    while(!gameOver())
    {
        playTurn();
        printScore();
        mTurnCounter++;
        std::cout << "\n";
    }
    std::cout << mTurnCounter<< " turns were played\n";
}


bool WarCardGame::gameOver() const
{
    if ((mP1Deck.empty() && mP1WinPile.empty()) ||
        (mP2Deck.empty() && mP2WinPile.empty()))
    {
        return true;
    }
    else
        return false;
}

unsigned long long WarCardGame::turnsPlayed() const
{
    return mTurnCounter;
}

/*
 * Recursive function if multiple wars occur
 *
 * For war, each player draws 2 cards. First card is "face down", aka just goes straight into
 * the warCards pile. The second is compared to decide who won.
 *
 * Edge cases to handle:
 *  - If player doesn't have enough cards to complete the war, they lose.
 *  - If neither do, the one who runs out first loses.
 *  - If both players run out simultaneously, it's a draw, they both get all their cards back.
 */
int WarCardGame::war()
{
    std::cout << "WAR!\n";
    
    replenishDecks(); // if war continues, need to replenish from win pile to continue the war.

    if (mP1Deck.empty() && mP2Deck.empty())
        return 3;
    else if (mP1Deck.empty())
    {
        std::cout << "Player 1 ran out of cards & could not continue the war\n";
        return 2;
    }
    else if (mP2Deck.empty())
    {
        std::cout << "Player 2 ran out of cards & could not continue the war\n";
        return 1;
    }

    mP1WarCards.push_back(mP1Deck.deal_card()); // face down cards
    mP2WarCards.push_back(mP2Deck.deal_card());

    if (mP1Deck.empty() && mP2Deck.empty())
        return 3;
    else if (mP1Deck.empty())
    {
        std::cout << "Player 1 ran out of cards & could not continue the war\n";
        return 2;
    }
    else if (mP2Deck.empty())
    {
        std::cout << "Player 2 ran out of cards & could not continue the war\n";
        return 1;
    }

    const PlayingCard card1 = mP1Deck.deal_card(); // face up cards to compare
    const PlayingCard card2 = mP2Deck.deal_card();
    mP1WarCards.push_back(card1);
    mP2WarCards.push_back(card2);

    std::cout << "Player 1 shows: " << card1.str() << "  Player 2 shows: " << card2.str() << std::endl;
    if (card1.rank() == card2.rank())
        return war(); // recursive
    else if (card1.rank() == PlayingCard::Rank::Ace || card1.rank() > card2.rank())
        return 1;
    else if (card2.rank() == PlayingCard::Rank::Ace)
        return 2;
    else if (card1.rank() < card2.rank())
        return 2;
    else
        return 1;
}

void WarCardGame::replenishDecks()
{
    if (mP1Deck.empty())
    {
        mP1Deck.assign(mP1WinPile);
        mP1WinPile.clear();
        mP1Deck.shuffle();
    }

    if (mP2Deck.empty())
    {
        mP2Deck.assign(mP2WinPile);
        mP2WinPile.clear();
        mP2Deck.shuffle();
    }
}

void WarCardGame::initialize(const std::vector<PlayingCard>& cards)
{
    Deck<PlayingCard> dealDeck(cards);
    dealDeck.shuffle();
    mP1WinPile.reserve(dealDeck.size()); // max these will ever be is total amount of cards
    mP2WinPile.reserve(dealDeck.size());
    while (!dealDeck.empty())
    {
        mP1WinPile.push_back(dealDeck.deal_card());
        mP2WinPile.push_back(dealDeck.deal_card());
    }

    mP1Deck.assign(mP1WinPile);
    mP2Deck.assign(mP2WinPile);
    mP1WinPile.clear();
    mP2WinPile.clear();
}

} // namespace doc


include/playing_card.h
/**
 * @file playing_card.h
 *
 * @author Channing Ko-Madden
 *
 * @date 2/18/2022
 *
 * @brief Declare PlayingCard class, that represents a French-suited playing card of a standard 52-card deck.
 *
 */

#ifndef PLAYING_CARD_H
#define PLAYING_CARD_H

#include <vector>
#include <string>

/**
 * @brief "Deck of Cards" namespace
 */
namespace doc
{

/**
 * @class PlayingCard "playing_card.h" "playing_card.h"
 * 
 * @brief The PlayingCard class represents a French-suited playing card of a standard 52-card deck. Cards can have
 * one four suites (clubs, diamonds, hearts, and spaces), and one of thirteen ranks (Ace, 2-10, Jack, Queen, 
 * King), also known as the face value.
 */
class PlayingCard
{
    public:

        /** @brief Represent the four suites */
        enum class Suite : int {Clubs = 0, Diamonds, Hearts, Spades};

        /** @brief Represent the thirteen ranks */
        enum class Rank : int {Ace = 1, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King};

        /** 
         * @brief Returns string representation of Suite enum
         * @param[in] suite
         * @return String version of suite
         */
        static std::string toString(const Suite suite);

        /** 
         * @brief Returns string representation of Rank enum
         * @param[in] rank 
         * @return String version of rank 
         */
        static std::string toString(const Rank rank);

        /**
         * @brief Return the underlying int value for a Rank enum.
         * Ace is low (Ace == 1)
         * @return Int value of Rank (Ace low)
         */
        static int toInt(const Rank rank);

        /**
         * @brief Construct playing card with suite and rank
         * @param[in] suite One of four suites
         * @param[in] rank One of thirteen ranks
         */
        PlayingCard(const Suite suite, const Rank rank);

        /**
         * @brief Returns the playing card's suite
         * @return suite
         */
        Suite suite() const;

        /**
         * @brief Returns the playing card's rank
         * @return rank 
         */
        Rank rank() const;

        /**
         * @brief Returns the card's suite and rank within a string.
         * Useful for logging or debugging
         */
        std::string str() const;

    private:
        Suite mSuite; /**<@brief Playing Card's suite */
        Rank mRank; /**<@brief Playing Card's rank */

};

std::vector<PlayingCard> buildStandardDeck();

} // namespace doc
#endif


include/war_card_game.h
/**
 * @file war_card_game.h
 *
 * @author Channing Ko-Madden
 *
 * @date 2/18/2022
 *
 * @brief Declare WarCardGame class, that simulates a game of the card game War, with the results
 * printed to standard out.
 *
 */

#ifndef WAR_CARD_GAME_H
#define WAR_CARD_GAME_H

#include "deck.h"
#include "playing_card.h"
#include <vector>

/**
 * @brief "Deck of Cards" namespace
 */
namespace doc
{

/**
 * @class WarCardGame "war_card_game.h" "war_card_game.h"
 * @brief This class simulates a game of the card game War, played with a standard 52 card deck, with the
 * results printed to standard out.
 *
 * War involves two players, and the goal is to be the first player to win all 52 cards.
 *
 * How to play:
 *  Each player draws a card and shows it. The higher card takes both cards and places them in a separate
 *  'winnings' pile. Ace is high.
 *
 *  If the cards are the same rank a WAR occurs, each player draws two additional cards,
 *  showing the second card drawn. The player with the higher rank card takes all six cards.
 *  If the shown cards are again the same rank, this process continues until there is a winner,
 *  with the winner taking all drawn cards.
 *
 *  When a player has run out of cards in their deck, they take their 'winnings' pile, shuffle it, and it becomes
 *  their deck to continue play with.
 *
 *  The game ends when one plyaer has won all the cards.
 */
class WarCardGame
{
    public:

        /**
         * @brief Constructor sets up standard 52 card game to play
         */
        WarCardGame();

        /**
         * @brief Construct game with custom assortment of playing cards
         *
         * @param[in] cards Cards to shuffle and deal to players
         * Must be an even number of cards for a fair game. If an odd number of cards is passed, the 
         * last card is dropped from the game.
         */
        WarCardGame(const std::vector<PlayingCard>& cards);

        /**
         * @brief Play a turn of the game
         */
        void playTurn();

        /**
         * @brief Prints the current game score to standard output.
         */
        void printScore() const;

        /**
         * @brief Plays the game until there is a winner, printing out the score after each turn, and the
         * total number of turns played.
         */
        void autoPlay();

        /**
         * @brief Returns if the game is over and there is a winner
         * @return True if the game is over, false if not.
         */
        bool gameOver() const;

        /**
         * @brief Returns the number of turns played so far.
         * A turn were one or more wars occur counts as one turn
         * @return Turns played
         */
        unsigned long long turnsPlayed() const;

    private:

        /**
         * @brief Initialize game with cards
         * @param[in] cards Cards to shuffle and deal out to players
         */
        void initialize(const std::vector<PlayingCard>& cards);

        /**
         * @brief Executes war play phase of game and returns the result of the war
         * Recursively calls itself if multiple wars occur
         *
         * The winner of war gets all the cards. If it is a draw, each player gets back
         * all their own cards.
         * @return 1 if Player One won, 2 if Player Two won, or 3 if it was a draw
         */
        int war();

        /**
         * @brief If either Player One's deck or Player Two's deck is empty, assign their win pile to
         * their decks and reshuffle the deck.
         *
         * If a deck is not empty, is it not modified.
         */
        void replenishDecks();

        Deck<PlayingCard> mP1Deck; /**<@brief Player One's deck to play from */
        Deck<PlayingCard> mP2Deck; /**<@brief Player Two's deck to play from */
        std::vector<PlayingCard> mP1WinPile; /**<@brief Player One's win pile, that becomes deck later */
        std::vector<PlayingCard> mP2WinPile; /**<@brief Player Two's win pile, that becomes deck later */
        std::vector<PlayingCard> mP1WarCards; /**<@brief Cards Player 1 has put down for war */
        std::vector<PlayingCard> mP2WarCards; /**<@brief Cards Player 2 has put down for war */
        unsigned long long mTurnCounter = 0; /**<@brief Counter for the number of turns played */

};

} // namespace doc
#endif
